view/src/extras/table.js

```diff
@@ -0,0 +1,105 @@
+import ArrowDownward from "@material-ui/icons/ArrowDownward";
+import Check from "@material-ui/icons/Check";
+import ChevronLeft from "@material-ui/icons/ChevronLeft";
+import ChevronRight from "@material-ui/icons/ChevronRight";
+import Clear from "@material-ui/icons/Clear";
+import DeleteOutline from "@material-ui/icons/DeleteOutline";
+import Edit from "@material-ui/icons/Edit";
+import FilterList from "@material-ui/icons/FilterList";
+import FirstPage from "@material-ui/icons/FirstPage";
+import LastPage from "@material-ui/icons/LastPage";
+import Remove from "@material-ui/icons/Remove";
+import SaveAlt from "@material-ui/icons/SaveAlt";
+import Search from "@material-ui/icons/Search";
+import ViewColumn from "@material-ui/icons/ViewColumn";
+
+/** Adds icons to all of the actions in the table */
+const tableIcons = {
+  Add: forwardRef((props, ref) => <AddBox {...props} ref={ref} />),
+  Check: forwardRef((props, ref) => <Check {...props} ref={ref} />),
+  Clear: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
+  Delete: forwardRef((props, ref) => <DeleteOutline {...props} ref={ref} />),
+  DetailPanel: forwardRef((props, ref) => (
+    <ChevronRight {...props} ref={ref} />
+  )),
+  Edit: forwardRef((props, ref) => <Edit {...props} ref={ref} />),
+  Export: forwardRef((props, ref) => <SaveAlt {...props} ref={ref} />),
+  Filter: forwardRef((props, ref) => <FilterList {...props} ref={ref} />),
+  FirstPage: forwardRef((props, ref) => <FirstPage {...props} ref={ref} />),
+  LastPage: forwardRef((props, ref) => <LastPage {...props} ref={ref} />),
+  NextPage: forwardRef((props, ref) => <ChevronRight {...props} ref={ref} />),
+  PreviousPage: forwardRef((props, ref) => (
+    <ChevronLeft {...props} ref={ref} />
+  )),
+  ResetSearch: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
+  Search: forwardRef((props, ref) => <Search {...props} ref={ref} />),
+  SortArrow: forwardRef((props, ref) => <ArrowDownward {...props} ref={ref} />),
+  ThirdStateCheck: forwardRef((props, ref) => <Remove {...props} ref={ref} />),
+  ViewColumn: forwardRef((props, ref) => <ViewColumn {...props} ref={ref} />),
+};
+
+/**
+ * Creates a custom table component that allows the user to search, edit,
+ * add, and delete entries into the table. The data that populates the
+ * table is loaded in through props passed by a parent.
+ */
+export default function CustomTable(props) {
+  // populates the table using the tableState property passed by the parent elem
+  const [tableState, setState] = React.useState(props.tableState);
+
+  return (
+    <MaterialTable
+      title={props.title}
+      // pulls the columns and data from the tableState state
+      columns={tableState.columns}
+      data={tableState.data}
+      // imports the icons
+      icons={tableIcons}
+      editable={{
+        // updates the table data after user adds row to table
+        onRowAdd: (newData) =>
+          new Promise((resolve) => {
+            setTimeout(() => {
```

Comment:
why do you need a setTimeout?

Reply:
I took this setup from the material-table documentation (I noticed it was weird too)

`https://material-table.com/#/docs/features/editable`

Reply:
It looks to me like that delay is to simulate the latency of a call to the server in their example.  You most definitely do not want that, since you will have real actual server latency instead.